File: src/main/java/io/cloudflight/architecture/structurizr/autoconfigure/StructurizrProperties.java
package io.cloudflight.architecture.structurizr.autoconfigure;

import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.bind.DefaultValue;

import java.io.File;

/**
 * Configuration properties for Structurizr.
 *
 * @author Klaus Lehner
 */
@ConfigurationProperties("structurizr")
public record StructurizrProperties(Workspace workspace, Export export) {

    public record Export(
            StructurizrExport structurizr,
            C4PlantUmlExport c4PlantUml
    ) {
    }

    /**
     * @param enabled                  if <code>true</code>, the workspace will be exported to a Structurizr server
     * @param url                      the url of the Structurizr server, the default is <code>https://api.structurizr.com</code>
     * @param id
     * @param key                      Workspace API key
     * @param secret                   Workspace API secret
     * @param encryptionPassphrase     Workspace client-side encryption passphrase
     * @param workspaceArchiveLocation Directory to use to archive workspaces.
     */
    public record StructurizrExport(
            @DefaultValue("false")
            boolean enabled,
            @DefaultValue("https://api.structurizr.com")
            String url,
            @DefaultValue("0")
            Long id,
            String key,
            String secret,
            String encryptionPassphrase,
            File workspaceArchiveLocation
    ) {
    }

    /**
     * @param enabled         if <code>true</code>, the Workspace will be exported using the PlantUML exporter
     * @param outputDirectory
     */
    public record C4PlantUmlExport(
            @DefaultValue("false")
            boolean enabled,
            @DefaultValue("build/c4PlantUml")
            File outputDirectory
    ) {
    }

    /**
     * @param name                       workspace name
     * @param description                workspace description
     * @param isAddImplicitRelationships Propagates all relationships from children to their parents.
     */
    public record Workspace(
            String name,
            String description,
            @DefaultValue("true")
            boolean isAddImplicitRelationships
    ) {
    }
}

File: src/main/kotlin/io/cloudflight/architecture/structurizr/autoconfigure/StructurizrProperties.kt
package io.cloudflight.architecture.structurizr.autoconfigure

import org.springframework.boot.context.properties.ConfigurationProperties
import java.io.File

/**
 * Configuration properties for Structurizr.
 *
 * @author Klaus Lehner
 */
@ConfigurationProperties("structurizr")
class StructurizrProperties {
    /**
     * Object containing all properties of your workspace at Structurizr
     */
    var workspace: Workspace = Workspace()
    var export: Export = Export()

    class Export {
        var structurizr: StructurizrExport = StructurizrExport()
        var c4PlantUml: C4PlantUmlExport = C4PlantUmlExport()
    }

    class StructurizrExport {
        /**
         * if true, the workspace will be exported to a Structurizr server
         */
        var enabled = false

        /**
         * The URL of Structurizr.
         */
        var url = "https://api.structurizr.com"

        /**
         * Workspace identifier.
         */
        var id: Long = 0

        /**
         * Workspace API key.
         */
        var key: String? = null

        /**
         * Workspace API secret.
         */
        var secret: String? = null

        /**
         * Workspace client-side encryption passphrase
         */
        var encryptionPassphrase: String? = null

        /**
         * Directory to use to archive workspaces.
         */
        var workspaceArchiveLocation: File? = null
    }

    class C4PlantUmlExport {
        /**
         * if true, the Workspace will be exported using the PlantUML exporter
         */
        var enabled = false
        var outputDirectory = File("build/c4PlantUml")
    }

    class Workspace {
        /**
         * Workspace name.
         */
        var name: String? = null

        /**
         * Workspace description.
         */
        var description: String? = null

        /**
         * Propagates all relationships from children to their parents.
         */
        var isAddImplicitRelationships = true
    }
}